### ISSUE TITLE
Kronecker-factored Approximate Curvature (K-FAC)

### ISSUE THREAD

**Issue by LeeEdel** [Lasagne/Lasagne#694]
http://arxiv.org/abs/1503.05671

Any chance someone could implement this for lasagne.updates?
It could be very good as it overcomes some of the issues with HF while also being faster to converge  than sgd with momentum.

Matlab code and additional information is provided here (http://www.cs.toronto.edu/~jmartens/research.html).

**Comment by benanne** [comment]
I'm not sure how feasible it is to implement this as a function in lasagne.updates, to be honest. I'm not super familiar with the method, but doesn't it make extensive use of the architectural properties of the network? I think it requires the network to consist of a stack of (non-convolutional) layers, for example. But functions in lasagne.updates should ideally be agnostic to the precise network architecture. All they see is a bunch of parameter variables, they are not aware of the internal structure.

Of course a Lasagne implementation of this would be great, but unfortunately I suspect it will be a bit more involved than simply adding another function to lasagne.updates.

**Comment by LeeEdel** [comment]
K-FAC has gotten some updates and new press:

https://jimmylba.github.io/papers/nsync.pdf

https://www.inboxdb.com/import-ai-cheaper-neural-network-training-mysterious-claims-around-bayesian-program/

"Half-price neural networks thanks to algorithmic tweaks: new research, Distributed Second-Order Optimization Using Kronecker-Factored Approximations (PDF), creates a new optimization method for training AI systems. The approach is flexible and can be dropped into pre-existing software relatively easily, its creators say. Best of all? “We show that our distributed K-FAC method speeds up training of various state-of-the-art ImageNet classification models by a factor of two compared to an improved form of Batch Normalization”. Quite rare to wake up one day and discover that your AI systems have just halved in price to train."

They also have a TensorFlow implementation available.

**Comment by f0k** [comment]
Probably the first step would be to try replicating some experiment of the paper as a Lasagne Recipe. This would already make it much easier to use it with Lasagne, and we can still add it to Lasagne afterwards if it can be implemented universally enough (and not architecture-dependent). Feel free to have a go!

**Comment by LeeEdel** [comment]
I'll probably try that out!

**Comment by wangxiaodong1021** [comment]
@LeeEdel Where can I find the code for the Tensorflow version?

**Comment by LeeEdel** [comment]
@wangxiaodong1021 I emailed the authors awhile ago about getting a copy of it. Last I heard they were cleaning the code up a bit before posting it.

**Comment by botev** [comment]
So, I've been working on something closely related to K-FAC. However, I don't think that you can hope to implement this without:
a) Implementing on Theano level, rather than lasagne
b) Keeping track of lots of intermediary variables along the graph of the network.
I think since James Martens joined DeepMind since September, they will be implementing this on Tensorflow level, e.g. a). 
Note that this is an approximate method, thus "standard" autodiff can not calculate everything for you.

**Comment by f0k** [comment]
> it essentially requires to have a module based `forward` and `backward` similar to torch/pytorch

Maybe `OpFromGraph` with `inline=True` and a custom `grad_overrides` function can help there? It allows you to define a module with a custom forward and backward pass given in terms of Theano expressions.

**Comment by botev** [comment]
Probably that would be helpful, but doesn't that mean that the user will have to rewrite all of his code to wtv ops you make with that, as well as there are extra global parameters to KFAC, which you must dynamically or globally set.

Additionally, you will need to do this for every single operator. Could you give a small example of how to use this, as I have no idea and I don't think there is a good example of it? For instance, consider instead of gradients I want to return gradients times a constant which you set somehow globally.

**Comment by f0k** [comment]
> How to bootstrap Lasagne, such that suddenly for all the implemented ops to use this new ops rather than the normal ones?

One could use a custom `lasagne.layers.get_output` routine that wraps the expressions returned by the layers.

> I think instead there has to be a new method on the Theano level, like Rop etc...

Hmm, but you'd want to implement this on a module/layer level, not on an Op level, right? That's the sole reason I suggested OpFromGraph, but as you need the original gradient as well, a separate extension that makes use of the Lasagne layer graph might be the easiest solution. Feel free to link your implementation here if you ever make it public!

**Comment by khaotik** [comment]
@botev 

So here's what I have in mind:

> How to use the OpFromGrad when you need to backpropagate 2 things

You could write an `OpFromGraph` with a copy of each original inputs. For example your Op has 5 inputs, then you can write an OfG with 10. For the forward pass, just use the first 5 inputs. For the backward pass, compute the first 5 grad as original grad and last 5 as custom grad.

When using the OfG instance, just give duplicated inputs. If you enabled `inline=True`, I think Theano *should* be able to eliminate redundant graph given all optimizations are on. (Not 100% sure tho)

```python
# defines custom op
def grad(inputs, grads):
    return [x_grad, y_grad, x_grad_custom, y_grad_custom]
op = OpFromGraph([x, y, x_copy, y_copy], [outputs, ...], grad_overrides=grad, inline=True)

# call with duplicated inputs
X_copy, Y_copy = X.copy(), Y.copy()
op(X, Y, X_copy, Y_copy)

# get gradients
X_grad, Y_grad, X_grad_custom, Y_grad_custom = T.grad(cost, [X, Y, X_copy, Y_copy])
```

Currently I'm quite new to Lasagne so no idea on that part.

**Comment by botev** [comment]
That indeed would be a neat solution, however, there are several caveats, which is why I did this in Lasagne and at least for the time being will keep it there:

1. It is hard to connect gradients of weight matrices correctly as you need to intercept whenever you have a shared variable, rather than a standard variable on the graph. 

2. Things like L2 decay can not be naively plugged in as they add a diagonal term to the factored matrix, rather than more gradients. 

3. You need a consistent notion of "samples" dimension.

4. You need a proper notion of what is a distribution, and that your loss is a log-likelihood.

5. For things like convolutional layers, it get's even more tricky. 

It is not impossible, but at the current stage, I don't have the time to spend on working on this, unfortunately.
